src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,386 @@
+package io.cdap.plugin.gcp.publisher.source;
+
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.auth.oauth2.ServiceAccountCredentials;
+import com.google.cloud.pubsub.v1.SubscriptionAdminClient;
+import com.google.cloud.pubsub.v1.stub.GrpcSubscriberStub;
+import com.google.cloud.pubsub.v1.stub.SubscriberStub;
+import com.google.cloud.pubsub.v1.stub.SubscriberStubSettings;
+import com.google.pubsub.v1.AcknowledgeRequest;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+import com.google.pubsub.v1.PushConfig;
+import com.google.pubsub.v1.ReceivedMessage;
+import org.apache.spark.storage.StorageLevel;
+import org.apache.spark.streaming.receiver.Receiver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.mutable.ArrayBuffer;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Spark Receiver for Pub/Sub Messages.
+ * <p>
+ * If backpressure is enabled, the message ingestion rate for this receiver will be managed by Spark.
+ */
+public class PubSubReceiver extends Receiver<ReceivedMessage> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PubSubReceiver.class);
+
+  protected BackoffConfig backoffConfig;
+  protected String project;
+  protected String topic;
+  protected String subscription;
+  protected boolean autoAcknowledge;
+  protected ServiceAccountCredentials credentials;
+
+  public PubSubReceiver(String project, @Nullable String topic, String subscription,
+                        ServiceAccountCredentials credentials, boolean autoAcknowledge, StorageLevel storageLevel) {
+    this(project, topic, subscription, credentials, autoAcknowledge, storageLevel,
+         BackoffConfigBuilder.getInstance().build());
+  }
+
+  public PubSubReceiver(String project, @Nullable String topic, String subscription,
+                        ServiceAccountCredentials credentials, boolean autoAcknowledge, StorageLevel storageLevel,
+                        BackoffConfig backoffConfig) {
+    super(storageLevel);
+
+    this.backoffConfig = backoffConfig;
+    this.project = project;
+    this.topic = topic;
+    this.subscription = subscription;
+    this.credentials = credentials;
+    this.autoAcknowledge = autoAcknowledge;
+  }
+
+  @Override
+  public void onStart() {
+    if (topic != null) {
```

Comment:
What does it mean when `topic == null`? Is it a valid use case?

Reply:
Yes, when the user supplies a Subscription name without a topic, we pull from the subscription (without attempting to create it).